src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/IProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,52 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using NBitcoin;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Consensus.ProvenBlockHeaders
+{
+    /// <summary>
+    /// Database of <see cref="ProvenBlockHeader"/>s.
+    /// </summary>
+    public interface IProvenBlockHeaderStore : IDisposable
+    {
+        /// <summary>
+        /// Loads <see cref="ProvenBlockHeader"/> items from the database.
+        /// </summary>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        Task InitializeAsync();
+
+        /// <summary>
+        /// Get a <see cref="ProvenBlockHeader"/> corresponding to a block.
+        /// </summary>
+        /// <param name="blockHeight">Height used to retrieve the <see cref="ProvenBlockHeader"/>.</param>
+        /// <returns><see cref="ProvenBlockHeader"/> retrieved from the <see cref="ProvenBlockHeaderStore"/>.</returns>
+        Task<ProvenBlockHeader> GetAsync(int blockHeight);
+
+        /// <summary>
+        /// Retrieves <see cref="ProvenBlockHeader"/> items.
+        /// </summary>
+        /// <param name="fromBlockHeight">Block height to start querying from.</param>
+        /// <param name="toBlockHeight">Block height to stop querying to.</param>
+        /// <returns>A dictionary of <see cref="ProvenBlockHeader"/> items by block height key.</returns>
+        Task<List<ProvenBlockHeader>> GetAsync(int fromBlockHeight, int toBlockHeight);
+
+        /// <summary>
+        /// Retrieves the <see cref="ProvenBlockHeader"/> of the current tip.
+        /// </summary>
+        /// <returns><see cref="ProvenBlockHeader"/></returns>
+        Task<ProvenBlockHeader> GetTipAsync();
+
+        /// <summary>
+        /// Adds a <see cref="ProvenBlockHeader"/> to the internal concurrent dictionary. 
+        /// </summary>
+        /// <param name="provenBlockHeader">A <see cref="ProvenBlockHeader"/> item to add.</param>
+        /// <param name="newTip">Block hash and height pair associated against the <see cref="ProvenBlockHeader"/> item.</param>
+        void AddToPending(ProvenBlockHeader provenBlockHeader, HashHeightPair newTip);       
```

Comment:
cache and batch should be separated. Otherwise we are facing same problem we have with coinview- when there are too many dirty items we don't have enough space for cache.

Reply:
Renamed to `AddToPendingBatch()` after a discussion with @dangershony .